src/bind.cpp

```diff
@@ -655,11 +643,29 @@ static std::unique_ptr<Expr> rebind_match(const std::string &fnname, ResolveBind
     assert (out.second);
   }
 
-  if (prototype.index == 1) {
-    prototype.tree = std::move(children.front());
+  std::shared_ptr<Sum> multiarg;
+  if (match->args.size() == 1) {
+    std::unique_ptr<Expr> arg = std::move(match->args.front());
+    prototype.tree.location = arg->location;
+    prototype.tree.var = 0;
+    for (auto &p : match->patterns)
+      arg = ascribe(std::move(arg), p.pattern.region, std::move(p.pattern.type));
+    auto out = map->defs.insert(std::make_pair("_ a0", DefValue(LOCATION, std::move(arg))));
+    assert (out.second);
   } else {
-    prototype.tree.children = std::move(children);
-    prototype.tree.sum = multiarg;
+    prototype.tree.sum = multiarg = std::make_shared<Sum>(AST(LOCATION));
+    multiarg->members.emplace_back(AST(LOCATION));
+    size_t i = 0;
+    for (auto &a : match->args) {
+      prototype.tree.children.emplace_back(a->location, i);
+      for (auto &p : match->patterns)
+        if (i < p.pattern.args.size())
+          a = ascribe(std::move(a), p.pattern.args[i].region, std::move(p.pattern.args[i].type));
```

Comment:
The reorganization of this function was entirely to get this line (663) distinct from the other line (652) for the cases of single-argument match and multi-argument match.